src/couch_stats/src/csrt_server.erl

```diff
@@ -0,0 +1,196 @@
+
+-spec set_context_type(Type, PidRef) -> boolean() when
+    Type :: rctx_type(), PidRef :: maybe_pid_ref().
+set_context_type(Type, PidRef) ->
+    update_element(PidRef, [{#rctx.type, Type}]).
+
+-spec create_resource(Rctx :: rctx()) -> true.
+create_resource(#rctx{} = Rctx) ->
+    catch ets:insert(?MODULE, Rctx).
+
+-spec destroy_resource(PidRef :: maybe_pid_ref()) -> boolean().
+destroy_resource(undefined) ->
+    false;
+destroy_resource({_, _} = PidRef) ->
+    catch ets:delete(?MODULE, PidRef).
+
+-spec get_resource(PidRef :: maybe_pid_ref()) -> maybe_rctx().
+get_resource(undefined) ->
+    undefined;
+get_resource(PidRef) ->
+    catch case ets:lookup(?MODULE, PidRef) of
+        [#rctx{} = Rctx] ->
+            Rctx;
+        [] ->
+            undefined
+    end.
+
+-spec is_rctx_field(Field :: rctx_field() | atom()) -> boolean().
+is_rctx_field(Field) ->
+    maps:is_key(Field, ?KEYS_TO_FIELDS).
+
+-spec get_rctx_field(Field :: rctx_field()) -> non_neg_integer().
+get_rctx_field(Field) ->
+    maps:get(Field, ?KEYS_TO_FIELDS).
+
+-spec update_counter(PidRef, Field, Count) -> non_neg_integer() when
+    PidRef :: maybe_pid_ref(),
+    Field :: rctx_field(),
+    Count :: non_neg_integer().
+update_counter(undefined, _Field, _Count) ->
+    0;
+update_counter({_Pid, _Ref} = PidRef, Field, Count) when Count >= 0 ->
+    %% TODO: mem3 crashes without catch, why do we lose the stats table?
+    case is_rctx_field(Field) of
+        true ->
+            Update = {get_rctx_field(Field), Count},
+            catch ets:update_counter(?MODULE, PidRef, Update, #rctx{pid_ref = PidRef});
+        false ->
+            0
+    end.
+
+-spec inc(PidRef :: maybe_pid_ref(), Field :: rctx_field()) -> non_neg_integer().
+inc(PidRef, Field) ->
+    inc(PidRef, Field, 1).
+
+-spec inc(PidRef, Field, N) -> non_neg_integer() when
+    PidRef :: maybe_pid_ref(),
+    Field :: rctx_field(),
+    N :: non_neg_integer().
+inc(undefined, _Field, _) ->
+    0;
+inc(_PidRef, _Field, 0) ->
+    0;
+inc({_Pid, _Ref} = PidRef, Field, N) when is_integer(N) andalso N >= 0 ->
+    case is_rctx_field(Field) of
+        true ->
+            update_counter(PidRef, Field, N);
+        false ->
+            0
+    end.
+
+%%
+%% gen_server callbacks
+%%
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+init([]) ->
+    ets:new(?MODULE, [
+        named_table,
+        public,
+        {decentralized_counters, true},
+        {write_concurrency, true},
+        {read_concurrency, true},
+        {keypos, #rctx.pid_ref}
+    ]),
+    {ok, #st{}}.
+
+handle_call(_, _From, State) ->
+    {reply, ok, State}.
+
+handle_cast(_Msg, State) ->
+    {noreply, State, 0}.
+
+%%
+%% private functions
+%%
+
+-spec update_element(PidRef :: maybe_pid_ref(), Updates :: [tuple()]) -> boolean().
+update_element(undefined, _Update) ->
+    false;
+update_element({_Pid, _Ref} = PidRef, Update) ->
+    %% TODO: should we take any action when the update fails?
```

Comment:
At least convert `{EXIT, badarg}` to `false` to conform to spec (or update the spec).